File: ios/NativeDownloadHandler/NativeDownloadHandler.m
//
//  Copyright (c) Mendix, Inc. All rights reserved.
//
#import <Foundation/Foundation.h>
#import "NativeDownloadHandler.h"
#import <React/RCTEventEmitter.h>

@implementation NativeDownloadHandler

+ (NSString *) formatMessage:(NSString *)message {
  return [[[NativeDownloadHandler description] stringByAppendingString:@": "] stringByAppendingString:message];
}

- (id)init:(NSDictionary *) config doneCallback:(void (^)(void))doneCallback
      progressCallback:(void (^)(long long, long long)) progressCallback
          failCallback:(void (^)(NSError *err))failCallback {
  self.mimeType = config[@"mimeType"];
  self.connectionTimeout = [((NSNumber *) config[@"connectionTimeout"] ?: [NSNumber numberWithInt: 10000]) intValue] / 1000;
  self.doneCallback = doneCallback;
  self.progressCallback = progressCallback;
  self.failCallback = failCallback;
  return self;
}

- (void)download:(NSString *)urlString downloadPath:(NSString *)downloadPath {
  self.downloadPath = downloadPath;

  NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
  NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration
                                                        delegate:self
                                                   delegateQueue:nil];
  NSURL *url = [NSURL URLWithString:[urlString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]]];
  NSURLRequest *request = [NSURLRequest requestWithURL:url
                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                       timeoutInterval:self.connectionTimeout];
  [[session downloadTaskWithRequest:request] resume];
}

#pragma mark NSURLSession Delegate Methods

- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location {
  NSError *fileError;
  NSFileManager *manager = [NSFileManager defaultManager];
  
	if (self.mimeType != nil && downloadTask.response.MIMEType != self.mimeType) {
		NSError *error = [NSError errorWithDomain:NSArgumentDomain code:-1 userInfo:@{NSLocalizedDescriptionKey:@"MIME type not expected."}];
		self.failCallback(error);
  }

  if ([manager fileExistsAtPath:[self downloadPath]]) {
      NSError *error = [NSError errorWithDomain:NSURLErrorDomain code:-1 userInfo:@{NSLocalizedDescriptionKey:@"File already exists in the same path."}];
      self.failCallback(error);
      return;
  }

  NSURL *directoryUrl = [NSURL fileURLWithPath:[[self downloadPath] stringByDeletingLastPathComponent]];
  [manager createDirectoryAtURL:directoryUrl withIntermediateDirectories:YES attributes:nil error:&fileError];
  if (fileError != nil) {
    NSLog([NativeDownloadHandler formatMessage:@"Could not create path: %@"], fileError);
    self.failCallback(fileError);
    return;
  }

  NSURL *url = [NSURL fileURLWithPath:[self downloadPath]];
  [manager replaceItemAtURL:url withItemAtURL:location backupItemName:[[[self downloadPath] lastPathComponent] stringByAppendingString:@"_backup"] options:NSFileManagerItemReplacementUsingNewMetadataOnly resultingItemURL:nil error:&fileError];
  if (fileError != nil) {
  	[manager removeItemAtURL:url error:nil];
    NSLog([NativeDownloadHandler formatMessage:@"Could not copy path: %@"], fileError);
    self.failCallback(fileError);
    return;
  }

  NSLog(@"%@", [NativeDownloadHandler formatMessage:@"File saved successfully"]);
  self.doneCallback();
}

- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didWriteData:(int64_t)bytesWritten totalBytesWritten:(int64_t)totalBytesWritten totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
  NSLog(@"%@", [NSString localizedStringWithFormat:[NativeDownloadHandler formatMessage:@"Bytes written %d"], totalBytesWritten]);
  if (self.progressCallback != nil)
  	self.progressCallback(totalBytesWritten, totalBytesExpectedToWrite);
}

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionDownloadTask *)downloadTask didCompleteWithError:(NSError *)error {
  if (!error) return;

  NSLog([NativeDownloadHandler formatMessage:@"Could not download: %@"], error);
  self.failCallback(error);
}

@end

@implementation NativeDownloadModule

RCT_EXPORT_MODULE()

NSString *ERROR_DOWNLOAD_FAILED = @"ERROR_DOWNLOAD_FAILED";
NSString *DOWNLOAD_PROGRESS_EVENT = @"NativeDownloadModuleDownloadProgress";

RCT_EXPORT_METHOD(download:(NSString *)url downloadPath:(NSString *)downloadPath config:(NSDictionary *)config resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  @try {
    NativeDownloadHandler *handler = [[NativeDownloadHandler alloc] init:config doneCallback: ^() {
      resolve(nil);
    } progressCallback:^(long long received, long long total) {
      [self sendEventWithName: DOWNLOAD_PROGRESS_EVENT body: @{
        @"receivedBytes": [NSNumber numberWithLongLong: received],
        @"totalBytes": [NSNumber numberWithLongLong: total]}];
    } failCallback:^(NSError *error) {
      reject(ERROR_DOWNLOAD_FAILED, [NativeDownloadHandler formatMessage:[error localizedDescription]], error);
    }];
    [handler download:url downloadPath:downloadPath];
  } @catch(NSError *error) {
    return reject(ERROR_DOWNLOAD_FAILED, [NativeDownloadHandler formatMessage:@"Failed to download file"], error);
  }
}

- (NSArray<NSString *> *)supportedEvents
{
  return @[DOWNLOAD_PROGRESS_EVENT];
}

@end

File: ios/Modules/NativeDownloadHandler/NativeDownloadHandler.swift
import Foundation
import React

fileprivate func formatMessage(_ message: String) -> String {
    return "\(String(describing: NativeDownloadHandler.self)): \(message)"
}

class NativeDownloadHandler: NSObject {
    
    let mimeType: String?
    let connectionTimeout: Int
    let doneCallback: (() -> Void)?
    let progressCallback: ((Int64, Int64) -> Void)?
    let failCallback: ((Error) -> Void)?
    var downloadPath: String = ""
    
    init(_
         config: [String: Any],
         doneCallback: @escaping () -> Void,
         progressCallback: ((Int64, Int64) -> Void)?,
         failCallback: @escaping (Error) -> Void
    ) {
        if let timeoutValue = config["connectionTimeout"] as? NSNumber {
            self.connectionTimeout = timeoutValue.intValue / 1000
        } else {
            self.connectionTimeout = 10
        }
        self.mimeType = config["mimeType"] as? String
        self.doneCallback = doneCallback
        self.progressCallback = progressCallback
        self.failCallback = failCallback
        super.init()
    }
    
    func download(_ urlString: String, downloadPath: String) {
        self.downloadPath = downloadPath
        
        let configuration = URLSessionConfiguration.default
        let session = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
        
        guard let encodedUrlString = urlString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: encodedUrlString) else {
            let error = NSError(domain: NSURLErrorDomain, code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
            failCallback?(error)
            return
        }
        
        let request = URLRequest(url: url, cachePolicy: .useProtocolCachePolicy, timeoutInterval: TimeInterval(connectionTimeout))
        let downloadTask = session.downloadTask(with: request)
        downloadTask.resume()
    }
}

// MARK: - URLSessionDownloadDelegate
extension NativeDownloadHandler: URLSessionDownloadDelegate {
    
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
        let fileManager = FileManager.default
        
        // Check MIME type if specified
        if let expectedMimeType = mimeType,
           let responseMimeType = downloadTask.response?.mimeType,
           responseMimeType != expectedMimeType {
            let error = NSError(domain: UserDefaults.argumentDomain, code: -1, userInfo: [NSLocalizedDescriptionKey: "MIME type not expected."])
            failCallback?(error)
            return
        }
        
        // Check if file already exists
        if fileManager.fileExists(atPath: downloadPath) {
            let error = NSError(domain: NSURLErrorDomain, code: -1, userInfo: [NSLocalizedDescriptionKey: "File already exists in the same path."])
            failCallback?(error)
            return
        }
        
        // Create directory if needed
        let directoryUrl = URL(fileURLWithPath: (downloadPath as NSString).deletingLastPathComponent)
        do {
            try fileManager.createDirectory(
                at: directoryUrl,
                withIntermediateDirectories: true,
                attributes: nil
            )
        } catch {
            NSLog("%@", formatMessage("Could not create path: \(error)"))
            failCallback?(error)
            return
        }
        
        // Move downloaded file to final location
        let destinationUrl = URL(fileURLWithPath: downloadPath)
        let backupName = "\((downloadPath as NSString).lastPathComponent)_backup"
        
        do {
            _ = try fileManager.replaceItem(
                at: destinationUrl,
                withItemAt: location,
                backupItemName: backupName,
                options: .usingNewMetadataOnly,
                resultingItemURL: nil
            )
            NSLog("%@", formatMessage("File saved successfully"))
            doneCallback?()
        } catch {
            try? fileManager.removeItem(at: destinationUrl)
            NSLog("%@", formatMessage("Could not copy path: \(error)"))
            failCallback?(error)
        }
    }
    
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
        NSLog("%@", formatMessage("Bytes written \(totalBytesWritten)"))
        progressCallback?(totalBytesWritten, totalBytesExpectedToWrite)
    }
}

// MARK: - URLSessionTaskDelegate
extension NativeDownloadHandler: URLSessionTaskDelegate {
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard let error = error else { return }
        NSLog("%@", formatMessage("Could not download: \(error)"))
        failCallback?(error)
    }
}

// MARK: - React Native Bridge Module
@objcMembers
public class NativeDownloadModule: NSObject {
    
    private static let ERROR_DOWNLOAD_FAILED = "ERROR_DOWNLOAD_FAILED"
    
    public func download(
        _ url: String,
        downloadPath: String,
        config: [String: Any],
        onProgress: @escaping ([String: Int64]) -> Void,
        promise: Promise
    ) {
        
        let handler = NativeDownloadHandler(
            config,
            doneCallback: {
                promise.resolve(nil)
            },
            progressCallback: { received, total in
                onProgress(["receivedBytes": received, "totalBytes": total])
            },
            failCallback: { error in
                promise.reject(NativeDownloadModule.ERROR_DOWNLOAD_FAILED, formatMessage(error.localizedDescription), error)
            }
        )
        
        handler.download(url, downloadPath: downloadPath)
    }
}

//Checked
